market_making_examples/restingLimitMaker/index.ts

```diff
@@ -0,0 +1,484 @@
+	logger.info(`AsksPlaced: ${asksPlaced}`);
+
+	if (!dryRun) {
+		try {
+			const lookupTableAccount = await driftClient.fetchMarketLookupTableAccount();
+			const vTx = await driftClient.txSender.getVersionedTransaction(ixs, [lookupTableAccount], [], driftClient.opts);
+			const tx = await driftClient.txSender.sendVersionedTransaction(vTx, [], driftClient.opts);
+			logger.info(`Update orders tx (slot: ${tx.slot}): ${tx.txSig}`);
+		} catch (e) {
+			console.error(e);
+		}
+	}
+};
+
+const main = async () => {
+
+	const connection = new Connection(endpoint, {
+		wsEndpoint: wsEndpoint,
+		commitment: stateCommitment,
+	});
+
+	const driftClient = new DriftClient({
+		connection,
+		wallet,
+		programID: new PublicKey(sdkConfig.DRIFT_PROGRAM_ID),
+		opts: {
+			commitment: stateCommitment,
+			skipPreflight: false,
+			preflightCommitment: stateCommitment,
+		},
+		accountSubscription: {
+			type: 'websocket',
+		},
+		env: driftEnv,
+		txSenderConfig: {
+			type: 'retry',
+			timeout: 35000,
+		},
+		activeSubAccountId: subaccountId,
+		subAccountIds: [subaccountId],
+	});
+	await driftClient.subscribe();
+	await driftClient.fetchMarketLookupTableAccount();
+
+	const slotSubscriber = new SlotSubscriber(connection, {});
+	await slotSubscriber.subscribe();
+
+	const userMap = new UserMap(
+		driftClient,
+		driftClient.userAccountSubscriptionConfig,
+		false
+	);
+	await userMap.subscribe();
+
+	const dlobSubscriber = new DLOBSubscriber({
+		driftClient,
+		dlobSource: userMap,
+		slotSource: slotSubscriber,
+		updateFrequency: 1000,
+	});
+	await dlobSubscriber.subscribe();
+
+
+	const { currLeverage, accountUsdValue } = loadUserPortfolio(driftClient);
+	logger.info(`User account value: ${accountUsdValue}, current leverage: ${currLeverage}`);
+
+	// run the market making loop on each update
+	let updatInProgress = false;
+	dlobSubscriber.eventEmitter.on('update', async (dlob: DLOB) => {
+		if (updatInProgress) return;
+		updatInProgress = true;
+		const start = Date.now();
+		const currSlot = slotSubscriber.getSlot();
+		const oracleData = driftClient.getOracleDataForPerpMarket(perpMarketIndex);
+
+		const userAccountData = driftClient.getUserAccountAndSlot();
+		if (currSlot > userAccountData.slot + MAX_USER_ACCOUNT_STALE_SLOTS) {
+			logger.info(`User account data stale by ${currSlot - userAccountData.slot} slots, updating`);
+			const startFetch = Date.now();
+			await driftClient.getUser().fetchAccounts();
+			logger.info(`Fetch user accounts took ${Date.now() - startFetch}ms`);
+		}
+
+		const perpMarketAccount = driftClient.getPerpMarketAccount(perpMarketIndex)!;
+		const vAsk = calculateAskPrice(perpMarketAccount, oracleData);
+		const vBid = calculateBidPrice(perpMarketAccount, oracleData);
+		const vAskNum = convertToNumber(vAsk, PRICE_PRECISION);
+		const vBidNum = convertToNumber(vBid, PRICE_PRECISION);
+
+		const bestBidNode = (dlob.getMakerLimitBids(perpMarketIndex, currSlot, MarketType.PERP, oracleData).next().value as DLOBNode);
```

Comment:
maybe we should make getBestBid/getBestAsk equal to this? so it's more ergonomic?